uast.cpp

```diff
@@ -0,0 +1,119 @@
+#include "uast.h"
+#include "roles.h"
+
+#define GET(__NODE__, __PROP__, ...) (impl.__PROP__(__NODE__, ##__VA_ARGS__))
+
+#define check_ptr(__V__)                                     \
+  if (__V__ == NULL)                                         \
+  {                                                          \
+    throw std::runtime_error("missing node implementation"); \
```

Comment:
Be careful that this doesn't propagate to the C binding, I don't know what would happen but it surely won't be nice :)